HW2/P2/P2.py

```diff
@@ -0,0 +1,104 @@
+    print("Serial uncorrelated: {} seconds".format(t.interval))
+    serial_counts = counts.copy()
+
+    # fine grained
+    counts[:] = orig_counts
+    with Timer() as t:
+        move_data_fine_grained(counts, src, dest, 100)
+    assert counts.sum() == total, "Wrong total after move_data_fine_grained"
+    print("Fine grained uncorrelated: {} seconds".format(t.interval))
+
+    ########################################
+    # You should explore different values for the number of locks in the medium
+    # grained locking
+    ########################################
+    # A range of N values to test against
+    different_Ns = [2, 5, 10,20,50]
+    # Array to plot
+    uncorrelated_times = []
+    for N in different_Ns:
+        counts[:] = orig_counts
+        with Timer() as t:
+            move_data_medium_grained(counts, src, dest, 100, N)
+        assert counts.sum() == total, "Wrong total after move_data_medium_grained"
+        print("Medium grained uncorrelated: {} seconds".format(t.interval))
+        uncorrelated_times.append(t.interval)
+
+
+
+    ########################################
+    # Now use correlated data movement
+    ########################################
+    dest = src + np.random.randint(-10, 11, size=src.size)
+    dest[dest < 0] += 1000
+    dest[dest >= 1000] -= 1000
+    dest = dest.astype(np.int32)
+
+    # serial move
+    counts[:] = orig_counts
+    with Timer() as t:
+        move_data_serial(counts, src, dest, 100)
+    assert counts.sum() == total, "Wrong total after move_data_serial"
+    print("Serial correlated: {} seconds".format(t.interval))
+    serial_counts = counts.copy()
+
+    # fine grained
+    counts[:] = orig_counts
+    with Timer() as t:
+        move_data_fine_grained(counts, src, dest, 100)
+    assert counts.sum() == total, "Wrong total after move_data_fine_grained"
+    print("Fine grained correlated: {} seconds".format(t.interval))
+
+    ########################################
+    # You should explore different values for the number of locks in the medium
+    # grained locking
+    ########################################
+    # Array to plot
+    correlated_times = []
+    for N in different_Ns:
+        counts[:] = orig_counts
+        with Timer() as t:
+            move_data_medium_grained(counts, src, dest, 100, N)
+        assert counts.sum() == total, "Wrong total after move_data_medium_grained"
+        print("Medium grained correlated: {} seconds".format(t.interval))
+        correlated_times.append(t.interval)
+
+    #uncorrelated_times = [8.02018904686,7.98845601082,8.32344985008,8.58538603783,9.4727640152]
+
+    #correlated_times = [8.17925286293,8.28549385071,8.05273008347,8.22479701042,8.9023911953]
+    plt.plot(different_Ns, uncorrelated_times)
+    plt.plot(different_Ns, correlated_times)
+    plt.xlabel("N value")
+    plt.ylabel("Seconds")
+    plt.legend(['Medium uncorrelated times', 'Medium correlated times'])
+    plt.show()
+
```

Comment:
All the code in this problem looks good, very cleanly written, organized, and commented.